### PR TITLE
[FIX] sale_loyalty: fixed taxes not paid by gift card

### PR DESCRIPTION
**Pull request by agr-odoo**
Users may give gift cards to their customers. Acting like a payment
method, gift cards will lower the total amount of a sale order up to 0.
Currently, in case of fixed tax applied in the order, the tax amount
cannot be covered with the card balance

Steps to reproduce
- Generate a Gift Card of 100$
- Create a Sale Order with a product having:
  - price: 90$
  - quantity: 1
  - A fixed tax of 10$
- Add the gift card to the order

Issue: Fixed tax amount will be left uncovered.
This is an issue because when using the gift card (or anyhow having the total 0),
it means the fixed tax cost is supported by the merchant so it should be
covered by the gift card

opw-4422678

### CODE REVIEW

**Comment by agr-odoo** [comment]
**before patch**
screenshot_000

**after patch**
screenshot

File: addons/sale_loyalty/models/sale_order.py [hunk 1]
```diff
@@ -172,9 +172,14 @@ def _discountable_order(self, reward):
...
-            # non-fixed tax totals. This way fixed taxes will not be discounted
-            taxes = line.tax_id.filtered(lambda t: t.amount_type != 'fixed')
+            if reward.program_type == 'gift_card':
+                # In case of gift cards the order can be totally paid with the
+                # card balance
+                taxes = line.tax_id
+            else:
+                # To compute the discountable amount we get the subtotal and add
+                # non-fixed tax totals. This way fixed taxes will not be discounted
+                taxes = line.tax_id.filtered(lambda t: t.amount_type != 'fixed')
```
Comment (Demesmaeker) [inline addons/sale_loyalty/models/sale_order.py:182]: Please use the field `is_payment_program` in the program, to also cover eWallet =]

File: addons/sale_loyalty/tests/test_pay_with_gift_card.py [hunk 2]
```diff
@@ -231,3 +231,32 @@ def test_paying_with_gift_card_uses_gift_card_product_taxes(self):
         self.assertAlmostEqual(order.amount_total, before_gift_card_payment - 100.0)
         self.assertTrue(all(line.tax_id for line in order.order_line))
         self.assertEqual(gift_card_line.tax_id, self.tax_10pc_incl + self.tax_15pc_excl)
+
+    def test_paying_with_gift_card_fixed_tax(self):
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_pay_with_gift_card.py:235]:
```suggestion
    def test_paying_fixed_tax_with_gift_card(self):
```

File: addons/sale_loyalty/tests/test_pay_with_gift_card.py [hunk 3]
```diff
@@ -231,3 +231,32 @@ def test_paying_with_gift_card_uses_gift_card_product_taxes(self):
...
+            'amount': 10,
+        })
+        self.product_A.write({'list_price': 90})
+        self.product_A.taxes_id = tax_10_fixed
+
+        order = self.empty_order
+        order.write({'order_line': [
+            Command.create({
+                'product_id': self.product_A.id,
+                'name': 'Ordinary Product A',
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_pay_with_gift_card.py:255]: Nitpicking
```suggestion
                'name': "Ordinary Product A",
```

File: addons/sale_loyalty/tests/test_pay_with_gift_card.py [hunk 4]
```diff
@@ -231,3 +231,32 @@ def test_paying_with_gift_card_uses_gift_card_product_taxes(self):
...
+                'product_uom': self.uom_unit.id,
+                'product_uom_qty': 1.0,
+            })
+        ]})
+        self._apply_promo_code(order, gift_card.code)
+        order.action_confirm()
+        self.assertEqual(order.amount_total, 0, "Order is not totally paid")
```
Comment (Demesmaeker) [inline addons/sale_loyalty/tests/test_pay_with_gift_card.py:262]: Nitpicking
```suggestion
        self.assertEqual(order.amount_total, 0, "The order should be totally paid")
```

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello @agr-odoo 

Thanks for your fix, here are some comments before merging it; Have a nice day!

**Comment by agr-odoo** [comment]
@Demesmaeker 
Thank you for the review, all points have been addressed :+1:

**Review by Demesmaeker — APPROVED** [review]
Hello @agr-odoo 
Thanks for your change and the test!
@robodoo r+

**Pull request closed without merging**